### ISSUE TITLE
protoparse: add --descriptor_set_in option

### ISSUE THREAD

**Issue by mingzhi** [jhump/protoreflect#279]
As discussed in protobuf [issue #2725](https://github.com/protocolbuffers/protobuf/issues/2725), adding `--descriptor_set_in` to pipe pre-compiled protos directly into the parser is super helpful when using Bazel and other cases.

(FYI @lukesneeringer, we need this option inside Google to work with Bazel, otherwise we will need to use protoc instead -- we can discuss separately)

**Comment by jhump** [comment]
@mingzhi, I don't follow. The protoparse package is not a command-line tool. There are no flags.

What you want to do is simply _load_ the file descriptor set, instead of parsing any sources (there are functions already in the `github.com/jhump/protoreflect/desc` package for this).

If you want to see this stuff in use, take a look at `goprotoc`, which is a pure-Go re-implementation of protoc (though it does not include any builtin code-gen, it only executes plugins). It already implements this option, in fact: https://github.com/jhump/goprotoc/blob/master/app/goprotoc/options.go#L123

**Comment by mingzhi** [comment]
Sorry that I confused you. I was asking if we can add a field `DescriptorSetIn` in the [`protoparse.Parser`](https://github.com/jhump/protoreflect/blob/master/desc/protoparse/std_imports.go) struct for searching dependencies. We already have a `ImportPaths` field that the `Parser` can search for dependencies in files. It would be great if we can supply pre-compiled descriptors as well. In other words, we would like to have a pre-defined, user-supplied imports in descriptors, similar to the standard imports in [std_imports.go](https://github.com/jhump/protoreflect/blob/master/desc/protoparse/std_imports.go).

It is different to what `goprotoc` is offering, which assume that the input files to be parsed are already parsed and included in the file descriptors ([link](https://github.com/jhump/goprotoc/blob/f9ac52ead1f5681bca4aca4756e830ad84dbb814/app/goprotoc/goprotoc.go#L235)).

It is needed when the parser doesn't have access to the entire file system (on the cloud), but has pre-compiled dependencies, or that the parser doesn't need to re-compile the imports.

Thanks!

**Comment by mingzhi** [comment]
Maybe a field `ImportDescriptors` (comparable with `ImportPaths`) is more suitable.

**Comment by jhump** [comment]
@mingzhi, this still makes no sense to me. Then you're not using the `protoparse` package at all -- whose purpose is to _parse proto source code_ (as its name suggests).

The functionality you describe is easy for you to write using the APIs already available in this repo.

**Comment by mingzhi** [comment]
Sorry again -- I hope to see something like:
Given a `address.proto`
```protobuf
import "google/contact/person.proto";

message Address {
  google.contact.Person person = 1;
  string address = 2;
}
```
and a *compiled* protobuf for person (we don't have access to `person.proto` file), I like to parse the `address.proto`:
```
personDescriptorProto := readCompiledProto("person.proto.pb")
parser := protoparse.Parser{
  ImportDescriptors: []*dpb.FileDescriptorProto{personDescriptorProto},
}
parser.ParseFiles("address.proto")
```

Thanks!

**Comment by lukesneeringer** [comment]
Maybe I could help explain. :-)

Let's say you want to parse a proto. One of the things that proto does is import some _other_ proto. It might be a common one controlled by someone else (so similar to `google.protobuf.*`, but not one of those). It would be nice to be able to import the Go package of the published, compiled proto that is imported and depend on Go tooling to keep it up to date, rather than having to copy the proto over into your source tree.

This is what the protoreflect package already does with the `google/protobuf/*` protos -- it depends on `github.com/golang/protobuf` and uses the compiled descriptors. However, that functionality is not extensible.

I am interested in this also (@mingzhi and I are working on the same project) and am considering sending a PR for it. Would that be something you might consider?

**Comment by jhump** [comment]
@lukesneeringer, yes, I understood after @mingzhi's last comment. I was even mulling over starting on this the other day. I think it would be a pretty small change -- pull requests welcome, BTW!

**Comment by lukesneeringer** [comment]
Great! I may see if I can tackle this. It would make my own project's pull request smaller. :-)

**Comment by jhump** [comment]
Resolved by #284
